mlrun/model_monitoring/model_monitoring_batch.py

```diff
@@ -474,6 +475,120 @@ def calculate_inputs_statistics(
+        return self._get_datetime_field(
+            mlrun.common.schemas.model_monitoring.EventFieldType.FIRST_REQUEST
+        )
+
+    def _get_start_time(
+        self, query_start: datetime.datetime
+    ) -> Optional[datetime.datetime]:
+        start_time = self._get_first_request_time()
+        if not start_time:
+            logger.info("No requests have been made yet.")
+            return None
+        last_analyzed = self._get_last_analyzed_time()
+        if last_analyzed:
+            logger.info("Found latest analyzed time {}", start_time)
+            start_time = last_analyzed
+        elif self._has_last_analyzed_key():
+            logger.info(
+                "No latest analyzed time was found, probably as this is the first run "
+                "of the batch monitoring job. Starting at {}",
+                start_time,
+            )
+        else:
+            start_time = self._get_window_start_time(query_start)
+            logger.info(
+                "No latest analyzed time was found, probably as it's an old system. "
+                "Starting at {}",
+                start_time,
+            )
+        return start_time
+
+    def _get_window_end_time(self, start_time: datetime.datetime) -> datetime.datetime:
+        return start_time + self._window_len
+
+    def _get_window_start_time(self, end_time: datetime.datetime) -> datetime.datetime:
+        return end_time - self._window_len
+
+    def get_windows(
+        self, now_func: Callable[[], datetime.datetime]
+    ) -> Iterator[Tuple[datetime.datetime, datetime.datetime]]:
+        """Get batch window time ranges"""
+        query_start = now_func()
+        start_time = self._get_start_time(query_start)
+        if not start_time:
+            return None
+        end_time = self._get_window_end_time(start_time)
+
+        if end_time > query_start:
+            logger.info(
+                "The query window isn't over yet. Wait for the next time window."
+            )
+
+        while end_time <= query_start:
+            yield start_time, end_time
+            start_time = end_time
+            end_time = self._get_window_end_time(start_time)
+
+    @classmethod
+    def batch_dict2window_len(cls, batch_dict: dict) -> datetime.datetime:
+        # TODO: This will be removed in 1.5.0 once the job params can be parsed with different types
+        # Convert batch dict string into a dictionary
+        if isinstance(batch_dict, str):
+            cls._parse_batch_dict_str(batch_dict)
+
+        return datetime.timedelta(
+            minutes=batch_dict[
+                mlrun.common.schemas.model_monitoring.EventFieldType.MINUTES
+            ],
+            hours=batch_dict[
+                mlrun.common.schemas.model_monitoring.EventFieldType.HOURS
+            ],
+            days=batch_dict[mlrun.common.schemas.model_monitoring.EventFieldType.DAYS],
+        )
+
+    @staticmethod
+    def _parse_batch_dict_str(batch_dict: dict) -> None:
```

Comment:
I changed the type hint to `Union[dict, str]`.
The TODO comment below reads:
```py
        # TODO: This will be removed in 1.5.0 once the job params can be parsed with different types
        # Convert batch dict string into a dictionary
        if isinstance(batch_dict, str):
            cls._parse_batch_dict_str(batch_dict)
```

So I understood the `str` is going to be removed very soon and didn't include a test for it. What do you think?